fix(browser): Ensure pageload & navigation spans have correct data

mydea opened the pull request:
This PR fixes two things about our idle spans emitted from `browserTracingIntegration`:

1. The navigation name was sometimes incorrect. This is because we look at `window.location.pathname` at the time when the `popstate` event is emitted - but at this point, this may not be updated yet. So a `navigation` transaction would possibly have the pathname of the previous page as transaction name.
2. The request data is also possibly incorrect - this is set by `HttpContext` integration at event processing time. However, at this time the `window.location` data is also usually already of the following navigation, so the pageload would often have wrong request data associated to it. Now, we store this on the current scope at span creation time to ensure it is actually correct.

Lms24 commented on packages/browser/src/tracing/browserTracingIntegration.ts:
```diff
@@ -355,11 +355,21 @@ export const browserTracingIntegration = ((_options: Partial<BrowserTracingOptio
           sampled: spanIsSampled(idleSpan),
           dsc: getDynamicSamplingContextFromSpan(span),
         });
+
+        scope.setSDKProcessingMetadata({
+          normalizedRequest: undefined,
+        });
       },
     });
 
     setActiveIdleSpan(client, idleSpan);
 
+    // We store the normalized request data on the scope, so we get the request data at time of span creation
+    // otherwise, the URL etc. may already be of the following navigation, and we'd report the wrong URL
+    getCurrentScope().setSDKProcessingMetadata({
+      normalizedRequest: getHttpRequestData(),
+    });
+
```
m/q: maybe I'm missing some timing information but isn't there a good chance that setting the `normalizedRequest` to `undefined` in 360 applies to the same scope where we previously set it (370)?

mydea replied:
argh, actually that does not work I just noticed, it fixed one test but broke another one - as this resets the request before we even process the transaction 😬 need to do this in a different way... basically the problem was apparent in some e2e tests. I think it becomes problematic in certain cases, e.g. nextjs pages router, where routing instrumentation triggers navigation spans before the url is updated 🤔

Lms24 commented on packages/browser/src/tracing/browserTracingIntegration.ts:
```diff
@@ -459,16 +469,18 @@ export const browserTracingIntegration = ((_options: Partial<BrowserTracingOptio
 
-            if (from !== to) {
-              startingUrl = undefined;
+            startingUrl = undefined;
```
m/q: why did we remove the `from !== to` check?

mydea replied:
sorry, forgot to mention this - we actually have the same check in the history handler code, where we do not even trigger the handler in this case!

Lms24 replied:
nice, good catch!

Lms24 commented on packages/browser/src/tracing/browserTracingIntegration.ts:
```diff
@@ -459,16 +469,18 @@ export const browserTracingIntegration = ((_options: Partial<BrowserTracingOptio
-              startingUrl = undefined;
+            startingUrl = undefined;
+
+            // We wait a tick here to ensure that WINDOW.location.pathname is updated
+            setTimeout(() => {
```
m: I'm a bit worried about delaying the span start for a tick. I guess this comes down to the router or implementation but what if users synchronously push a new state and e.g. make a `fetch` call directly afterwards? Would we still catch the span?
i think alternatively, we could still start the span directly but call `updateSpanName` once more in the `setTimeout` closure.
Not ideal in terms of DSC consistency but I'd rather have the span than miss it. WDYT?

mydea commented:
OK, so I updated this based on feedback, to a more reliable heuristic I think:

1. Removed the `setTimeout` stuff. Instead, we now use `to` from the history instrumentation, which is correct even if the navigation is still pending.
2. However, to do this I had to fix the history instrumentation to ensure we always get a full URL - this was inconsistent before, where it could be a full URL or a path, depending on what is triggered. oops.
3. I updated it to only keep the `normalizedRequest` for the default browserTracingIntegration - other implementations of `startNavigationSpan` will not set this, keeping the old behavior. This is because some frameworks (e.g. nextjs pages router) may trigger this before the URL is available. We can look at this later still, if needed.

Lms24 reviewed (approved):
Thanks for addressing my feedback!
